pkg/apis/core/validation/validation.go

```diff
@@ -4446,11 +4456,52 @@ func ValidateServiceCreate(service *core.Service) field.ErrorList {
 func ValidateServiceUpdate(service, oldService *core.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
 
-	// ClusterIP should be immutable for services using it (every type other than ExternalName)
+	// clusterIPs[0] (and conditionally ipfamilies) must be immutable for services using it (every type other than ExternalName)
 	// which do not have ClusterIP assigned yet (empty string value)
 	if service.Spec.Type != core.ServiceTypeExternalName {
-		if oldService.Spec.Type != core.ServiceTypeExternalName && oldService.Spec.ClusterIP != "" {
-			allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
+		if oldService.Spec.Type != core.ServiceTypeExternalName && (len(oldService.Spec.ClusterIPs) > 0 && oldService.Spec.ClusterIPs[0] != "") {
+			// primary ip can not change, secondary ips can not change once set, but it can be removed
```

Comment:
It's useful to be able _add_ a secondary IP so you can upgrade your pre-existing single-stack services to dual-stack. I'm not sure there are any good use cases for removing the secondary IP other than "I screwed up and didn't mean to make this service dual-stack".

Reply:
ClusterIP (or ClusterIPs[0]) follows the same add/remove semantics via `ClusterIP=>ExternalName=>ClusterIP` service type conversion. We don't allow `change` but allow `release` .. `acquire new` (which is `change` via hoops jumping). The importance of add/remove 2nd ClusterIP comes from the need to upgrade/downgrade services to/from dual-stack. A n example use case is `there is no more need to expose a secondary IP family, the clients that needed it are no longer existent` (in addition to user error case @danwinship  described above :-))

I am all in for properly immutability across both IPs. Together at the same time. That is another big story not just api wise but also ClusterIP alloc/release management. And maybe across proxies as well. Can we handle one big story at a time please?

Reply:
No argument on the need to add 2nd.

If we are going to allow removing 2nd, what if we allow the user to flip them?  Hear me out...

Users can already bounce through ExternalName (or just delete and recreate) to change a ClusterIP.  We made it immutable a LONG time ago because we thought it was more pain to support it as mutable, but it's been a gadfly ever since.  If we allow the user to flip them, then they can do v4 -> v4,v6 -> v6,v4 -> v6. The effect of that 2nd change is, strictly, a breaking change.  I wonder if there's any way we can make that safe?  Uggh, thinking about it, maybe not.  

Clever ideas?  Kal - is the need to remove 2nd really real, or hypothetical?

Reply:
> An example use case is `there is no more need to expose a secondary IP family, the clients that needed it are no longer existent`

Is that plausible? It seems at least as likely, if not more likely, that you'd want to remove the _primary_ IP family. ie, you start with a v4 cluster, upgrade to dual-stack, make the services v4,v6, then eventually in the future get rid of all the v4 and make the services just v6.

People can always delete+recreate or cheat via ExternalName if they need to make some change we don't allow them to make directly...

Reply:
> An example use case is `there is no more need to expose a secondary IP family, the clients that needed it are no longer existent`

If it makes dual-stack easier to implement, my vote is the user is stuck with that second IP family until they delete + recreate the service with the actual options they want. Or via ExternalName.

For the user-error case, wouldn't a user have to delete/recreate if they messed up certain fields anyway?

Reply:
I really, really dislike the idea of oneway upgrade. Let us go back to reasons of doing it

We wanted to allow users to upgrade so they can preserve `clusterIPs[0]` on services. in other words users can make use of new features without having to lose investment they might have made in current api object instances. 

Similarly we should allow people to downgrade while preserving `ClusterIPs[0]` because locking users into a one way street is just bad and i would accept if the cost is not too high (in resources, not in complexities we can absorb in the code, even if we had to re write some pieces of this PR).

The point i am trying to make is. If we allowed the user to upgrade existing services, then we should allow downgrade.